proposals/route-blinding.md

```diff
@@ -0,0 +1,285 @@
+
+This scheme requires all the nodes in the blinded route and the sender to activate support for the
+feature. It only becomes effective once a big enough share of the network supports it.
+
+### Notations
+
+* A node `N(i)`'s `node_id` is defined as: `P(i) = k(i) * G`.
+* Blinded `node_id`s are defined as: `B(i) = b(i) * G`.
+* Ephemeral public keys are defined as: `E(i) = e(i) * G`.
+
+### Requirements
+
+A node `N(r)` wants to provide a blinded route `N(r) <- ... <- N(1) <- N(0)` that must be used
+to receive onion messages.
+
+* The channels used along that route may be either announced or unannounced.
+* When used for payments, intermediate nodes in the blinded route MUST NOT learn `payment_secret`.
+* Intermediate nodes in the blinded route MUST NOT learn the `node_id` or `scid` of other
+  intermediate nodes except for their immediate predecessor or successor.
+* Intermediate nodes in the blinded route MUST NOT learn their distance to the recipient `N(r)`.
+* Senders MUST NOT learn the real `node_id`s and `scid`s of the blinded intermediate hops after the
+  introduction point `N(0)`.
+* If `N(r)` creates multiple blinded routes to herself, senders MUST NOT be able to tell that these
+  routes lead to the same recipient (unless this information is leaked by higher layers of the
+  protocol, such as using the same `payment_hash`).
+
+### Encrypted recipient data
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_recipient_data`.
+
+This field is used to carry data coming from the final recipient that cannot be modified by the
+sender. For route blinding it only needs to contain the `scid` to use when forwarding the message,
+but it may be extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`encrypted_recipient_data`)
+2. data:
+    * [`...*byte`:`encrypted_recipient_data`]
+
+Once decrypted, the content of this encrypted payload is a TLV stream that may contain any tlv
+record defined in Bolt 4 (onion TLV namespace).
+
+### Creating a blinded route
+
+`N(r)` performs the following steps to create a blinded route:
+
+```text
+Initialization:
+
+  e(0) <- {0;1}^256
+  E(0) = e(0) * G
+
+Blinding:
+
+  For i = 0 to r-1:
+    ss(i) = H(e(i) * P(i)) = H(k(i) * E(i))         // shared secret known only by N(r) and N(i)
+    B(i) = HMAC256("blinded_node_id", ss(i)) * P(i) // Blinded node_id for N(i), private key known only by N(i)
+    rho(i) = HMAC256("rho", ss(i))                  // Key used to encrypt payload for N(i) by N(r)
+    e(i+1) = H(E(i) || ss(i)) * e(i)                // Ephemeral private key, only known by N(r)
+    E(i+1) = H(E(i) || ss(i)) * E(i)                // NB: N(i) must not learn e(i)
+
+Blinded route:
+  
+  (P(0),fees(0),cltv(0),encrypted_recipient_data(0))
+  (B(1),fees(1),cltv(1),encrypted_recipient_data(1))
+  ...
+  (B(r-1),fees(r-1),cltv(r-1),encrypted_recipient_data(r-1))
```

Comment:
It is unclear to me how the fees are encoded. Given that we want to allow the sender to vary the amounts involved we'd probably need to also communicate both `base_fee` and `proportional_fee` for each blinded hop, allowing the sender to compute them. So it's not a fixed number but rather a function of the amount the hop receives.

The same is true for the CLTV, but that's just a single delta.

Reply:
It's true, I kept some details quite vague for now: I wanted to make sure we agree on the scheme, crypto, etc before specifying those in depth.